subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectListTest.groovy

```diff
@@ -396,6 +396,31 @@ class DefaultNamedDomainObjectListTest extends AbstractNamedDomainObjectCollecti
         list.findAll { it != "b" } == ["a", "c"]
     }
 
+    def "name based filtering does not realize pending"() {
+        given:
+        list.add("realized1")
+        list.addLater(new TestNamedProvider("unrealized1", "unrealized1"))
+        list.add("realized2")
+        list.addLater(new TestNamedProvider("unrealized2", "unrealized2"))
+
+        expect: "unrealized elements remain as such"
+        list.index.asMap().size() == 2
+        list.index.pendingAsMap.size() == 2
+
+        when: "filter the list via the `named` method"
+        def filtered = list.named { it.contains("2") }
+
+        then: "unrealized elements remain as such"
+        list.index.asMap().size() == 2
+        list.index.pendingAsMap.size() == 2
+
+        filtered.index.asMap().size() == 1
+        filtered.index.pendingAsMap.size() == 1
+
+        and:
+        filtered.asList() == ["realized2", "unrealized2"]
```

Comment:
❌ this should assert the state of (un)realized elements after calling `toList()`